internal/uvm/scsi/manager.go

```diff
@@ -0,0 +1,322 @@
+package scsi
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/protocol/guestresource"
+	"github.com/Microsoft/hcsshim/internal/security"
+	"github.com/Microsoft/hcsshim/internal/verity"
+	"github.com/Microsoft/hcsshim/internal/wclayer"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// ErrNoAvailableLocation indicates that a new SCSI attachment failed because
+	// no new slots were available.
+	ErrNoAvailableLocation = errors.New("no available location")
+	// ErrNotInitialized is returned when a method is invoked on a nil [Manager].
+	ErrNotInitialized = errors.New("SCSI manager not initialized")
+	// ErrAlreadyReleased is returned when [Mount.Release] is called on a Mount
+	// that had already been released.
+	ErrAlreadyReleased = errors.New("mount was already released")
+)
+
+// Manager is the primary entrypoint for managing SCSI devices on a VM.
+// It tracks the state of what devices have been attached to the VM, and
+// mounted inside the guest OS.
+type Manager struct {
+	attachManager *attachManager
+	mountManager  *mountManager
+}
+
+// Slot represents a single SCSI slot, consisting of a controller and LUN.
+type Slot struct {
+	Controller uint
+	LUN        uint
+}
+
+// NewManager creates a new Manager using the provided backend [Attacher],
+// [Mounter], and [Unplugger], as well as other configuration parameters.
+//
+// guestMountFmt is the format string to use for mounts of SCSI devices in
+// the guest OS. It should have a single %d format parameter.
+//
+// reservedSlots indicates which SCSI slots to treat as already used. They
+// will not be handed out again by the Manager.
+func NewManager(
+	attacher Attacher,
+	mounter Mounter,
+	unplugger Unplugger,
+	numControllers int,
+	numLUNsPerController int,
+	guestMountFmt string,
+	reservedSlots []Slot,
+) *Manager {
+	am := newAttachManager(attacher, unplugger, numControllers, numLUNsPerController, reservedSlots)
+	mm := newMountManager(mounter, guestMountFmt)
+	return &Manager{am, mm}
+}
+
+// MountConfig specifies the options to apply for mounting a SCSI device in
+// the guest OS.
```

Comment:
It feels weird to me that the way we decide if we should mount a device in the guest is based on if we pass through a MountConfig. Considering the fact that the only two fields on the MountConfig are "Encrypted" and "Options", having a nil MountConfig seems to indicate that neither of those two fields have content or that we aren't configuring the mount in a unique way, not that we should avoid mounting all together. Thoughts?

Reply:
I was distinguishing between a nil `MountConfig` (do not mount) vs a zero value `MountConfig` (mount but without any special options). There are cases where we want to attach without mounting (such as in `runhcs create-scratch`), so we need some way to distinguish.

My other thought was we could expose `attachManager` and `mountManager` externally, and have clients call into them individually to set up attachment vs mount. This may be what we want to do long-term anyways (since a mount is not a SCSI-specific concept), but it seemed like a fair increase in verbosity, so I tried to find a way to keep it as one call for now.

Reply:
Yeah, I agree keeping the attach and mount in a single call for the external caller is probably fine/best for now. However, instead of changing to expose those managers for now, could we add a new call to the MountManager that just attaches the disk, but doesn't mount it? What do you think?

Reply:
We would need to add 3 new methods, for each of VirtualDisk, PhysicalDisk, and ExtensibleVirtualDisk. Do you think passing a nil MountConfig to indicate "don't mount" is too confusing and thus warrants this?

Reply:
Resolving this since we're discussing more about it in [this thread](https://github.com/microsoft/hcsshim/pull/1744#discussion_r1175904281).